lib/client/client.go

```diff
@@ -49,40 +90,35 @@ type client struct {
 	callOptions []grpc.CallOption
 }
 
-// NewClientFunc defines a function that returns RPC agent client.
-type NewClientFunc func(ctx context.Context, addr, caFile, certFile, keyFile string) (*client, error)
-
 // NewClient creates a agent RPC client to the given address
 // using the specified client certificate certFile
-func NewClient(ctx context.Context, addr, caFile, certFile, keyFile string) (*client, error) {
-	return newClient(ctx, addr, "", caFile, certFile, keyFile)
-}
+func NewClient(ctx context.Context, config Config) (*client, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-// newClient additional allows a serverName override, but is only available
-// from unit tests
-func newClient(ctx context.Context, addr, serverName, caFile, certFile, keyFile string) (*client, error) {
 	// Load client cert/key
-	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	cert, err := tls.LoadX509KeyPair(config.CertFile, config.KeyFile)
 	if err != nil {
 		return nil, trace.ConvertSystemError(err)
 	}
 
 	// Load the CA of the server
-	clientCACert, err := ioutil.ReadFile(caFile)
+	clientCACert, err := ioutil.ReadFile(config.CAFile)
 	if err != nil {
 		return nil, trace.ConvertSystemError(err)
 	}
 
 	certPool := x509.NewCertPool()
 	if !certPool.AppendCertsFromPEM(clientCACert) {
-		return nil, trace.Wrap(err, "failed to append certificates from %v", caFile)
+		return nil, trace.Wrap(err, "failed to append certificates from %v", config.CAFile)
 	}
 
 	creds := credentials.NewTLS(&tls.Config{
 		RootCAs:      certPool,
 		Certificates: []tls.Certificate{cert},
 		MinVersion:   tls.VersionTLS12,
-		ServerName:   serverName,
+		ServerName:   "",
```

Comment:
Why this change?

Reply:
The option to override `serverName` was only available for unit tests. In normal use `NewClient` was initializing client with no `ServerName`.
Moved this functionality out of `client.go` and into the test file 'client_test.go` to simplify the code.